keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2150 @@
+	// those stored in ResourceClaimSpec. These parameters here can only be
+	// set by cluster administrators.
+	//
+	// The object must be cluster-scoped.
+	Parameters TypedLocalObjectReference
+}
+
+// ResourceClaim is created by users to describe which resources they need.
+// Its status tracks whether the resource has been allocated and what the
+// resulting attributes are.
+type ResourceClaim struct {
+	metav1.TypeMeta
+
+	// The driver must set a finalizer here before it attempts to allocate
+	// the resource. It removes the finalizer again when a) the allocation
+	// attempt has definitely failed or b) when the allocated resource was
+	// freed. This ensures that resources are not leaked.
+	//
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ObjectMeta
+
+	// Spec describes the desired attributes of a resource that then needs
+	// to be allocated. It can only be set once when creating the
+	// ResourceClaim.
+	Spec core.ResourceClaimSpec
+
+	// Status describes whether the resource is available and with which
+	// attributes.
+	Status ResourceClaimStatus
+}
+
+// ResourceClaimSpec defines how a resource is to be allocated.
+type ResourceClaimSpec struct {
+	// ResourceClassName references the driver and additional parameters
+	// via the name of a ResourceClass that was created as part of the
+	// driver deployment.
+	//
+	// The apiserver does not check that the referenced class exists, but a
+	// driver-specific admission webhook may require that and is allowed to
+	// reject claims where the class is missing.
+	ResourceClassName string
+
+	// Parameters references a separate object with arbitrary parameters
+	// that will be used by the
+	// driver when allocating a resource for the claim.
+	//
+	// The object must be in the same namespace as the ResourceClaim.
+	Parameters TypedLocalObjectReference
+
+	// Allocation can start immediately or when a Pod wants to use the
+	// resource. Waiting for a Pod is the default.
+	AllocationMode AllocationMode
+}
+
+// AllocationMode describes whether a ResourceClaim gets allocated immediately
+// when it gets created (AllocationModeImmediate) or whether allocation is
+// delayed until it is needed for a Pod (AllocationModeDelayed). Other modes
+// might get added in the future.
+type AllocationMode string
+
+const (
+	// When a ResourceClaim has AllocationModeDelayed, allocation is
+	// delayed until a Pod gets scheduled that needs the ResourceClaim. The
+	// scheduler will consider all resource requirements of that Pod and
+	// trigger allocation for a node that fits the Pod.
+	AllocationModeDelayed AllocationMode = "Delayed"
+
+	// When a ResourceClaim has AllocationModeImmediate, allocation starts
+	// as soon as the ResourceClaim gets created. This is done without
+	// considering the needs of Pods that will use the ResourceClaim
+	// because those Pods are not known yet.
+	AllocationModeImmediate AllocationMode = "Immediate"
```

Comment:
Why not use the same constants as for volumes? `Immediate` and `WaitForFirstConsumer`?

Reply:
Because `WaitForFirstConsumer` is such an awkward word. In documentation, we often talk about "delayed provisioning" and then have to explain that this is chosen by setting `WaitForFirstConsumer`. 
"Immediate" and "Delayed" look more like opposites.

But I don't feel strongly about this. I can see how consistency with volumes would be useful.

Reply:
That may have been the reason for using `WaitForFirstConsumer`. Let's stick with it for now in the KEP and I'll add a TODO in my prototype so that we don't forget about it.